eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/NodeRelay.scala

```diff
@@ -125,15 +126,32 @@ object NodeRelay {
       Some(InvalidOnionPayload(UInt64(2), 0))
     } else {
       payloadOut match {
-        case payloadOut: IntermediatePayload.NodeRelay.Standard =>
-          if (payloadOut.invoiceFeatures.isDefined && payloadOut.paymentSecret.isEmpty) {
-            Some(InvalidOnionPayload(UInt64(8), 0)) // payment secret field is missing
-          } else {
-            None
-          }
-        case _: IntermediatePayload.NodeRelay.ToBlindedPaths =>
-          None
+        // If we're relaying a standard payment to a non-trampoline recipient, we need the payment secret.
+        case payloadOut: IntermediatePayload.NodeRelay.Standard if payloadOut.invoiceFeatures.isDefined && payloadOut.paymentSecret.isEmpty => Some(InvalidOnionPayload(UInt64(8), 0))
+        case _: IntermediatePayload.NodeRelay.Standard => None
+        case _: IntermediatePayload.NodeRelay.ToBlindedPaths => None
+      }
+    }
+  }
+
+  private def nextNodeIsWallet(nodeParams: NodeParams, recipient: Recipient): Boolean = {
+    recipient match {
+      case r: BlindedRecipient => r.blindedHops.head.resolved.route match {
+        case BlindedPathsResolver.PartialBlindedRoute(_: EncodedNodeId.WithPublicKey.Wallet, _, _) => true
+        case _ => false
       }
+      case _ => false
+    }
+  }
+
+  /** When we have identified that the next node is one of our peers, return their (real) nodeId. */
+  private def walletNodeId(recipient: Recipient): PublicKey = {
```

Comment:
This function doesn't really make sense on its own, it should be combined with `nextNodeIsWallet`

```scala
private def nextWalletNodeId(nodeParams: NodeParams, recipient: Recipient): Option[PublicKey] = {
  recipient match {
    case r: BlindedRecipient => r.blindedHops.head.resolved.route match {
      case BlindedPathsResolver.PartialBlindedRoute(EncodedNodeId.WithPublicKey.Wallet(publicKey), _, _) =>
        Some(publicKey)
      case _ => None
    }
    case _ => None
  }
}
```

Reply:
Good point, refactored in https://github.com/ACINQ/eclair/pull/2865/commits/fa5b4376f561fe35bb78e6a19675ae129f31b194